src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,442 @@
+ * to come to our rescue. But sql (not plpgsql) functions unfortunately don't
+ * go through a codepath supporting custom plans. We call VerifyMultiPlanValidity()
+ * to do this check and provide a meaningfull error message.
  */
-static void
-CopyQueryResults(List *masterCopyStmtList)
+Node *
+InvalidCreateScan(CustomScan *scan)
 {
-	ListCell *masterCopyStmtCell = NULL;
+	CitusScanState *scanState = palloc0(sizeof(CitusScanState));
 
-	/* now copy data from all the remote nodes into temp table */
-	foreach(masterCopyStmtCell, masterCopyStmtList)
-	{
-		Node *masterCopyStmt = (Node *) lfirst(masterCopyStmtCell);
+	scanState->executorType = MULTI_EXECUTOR_INVALID_FIRST;
+	scanState->customScanState.ss.ps.type = T_CustomScanState;
+	scanState->multiPlan = GetMultiPlan(scan);
 
-		Assert(IsA(masterCopyStmt, CopyStmt));
+	Assert(IsA(scanState, CustomScanState));
 
-		ProcessUtility(masterCopyStmt,
-					   "(copy job)",
-					   PROCESS_UTILITY_QUERY,
-					   NULL,
-					   None_Receiver,
-					   NULL);
-	}
+	/* ensure plan is executable */
+	VerifyMultiPlanValidity(scanState->multiPlan);
 
-	/* make the copied contents visible */
-	CommandCounterIncrement();
+	return (Node *) scanState;
 }
 
 
-/* Execute query plan. */
+/*
+ * CitusSelectBeginScan just checks if the given custom scan node is a proper
+ * Citus scan node.
+ */
 void
-multi_ExecutorRun(QueryDesc *queryDesc, ScanDirection direction, tuplecount_t count)
+CitusSelectBeginScan(CustomScanState *node, EState *estate, int eflags)
 {
-	int eflags = queryDesc->estate->es_top_eflags;
-
-	if (eflags & EXEC_FLAG_CITUS_ROUTER_EXECUTOR)
-	{
-		/* drop into the router executor */
-		RouterExecutorRun(queryDesc, direction, count);
-	}
-	else
-	{
-		/* drop into the standard executor */
-		standard_ExecutorRun(queryDesc, direction, count);
-	}
+	ValidateCitusScanState(node);
 }
 
 
-/* Perform actions, like e.g. firing triggers, after the query has run. */
+/*
+ * ValidateCitusScanState checks if the given scan nodes contains a valid multi plan.
+ */
 void
-multi_ExecutorFinish(QueryDesc *queryDesc)
+ValidateCitusScanState(CustomScanState *node)
 {
-	int eflags = queryDesc->estate->es_top_eflags;
+	CitusScanState *scanState = (CitusScanState *) node;
+	MultiPlan *multiPlan = scanState->multiPlan;
 
-	if (eflags & EXEC_FLAG_CITUS_ROUTER_EXECUTOR)
+	Assert(IsA(scanState, CustomScanState));
+
+	/* ensure plan is executable */
+	VerifyMultiPlanValidity(multiPlan);
+}
+
+
+/*
+ * RealTimeExecScan is a callback function which returns next tuple from a real-time
+ * execution. In the first call, it executes distributed real-time plan and loads
+ * results from temporary files into custom scan's tuple store. Then, it returns
+ * tuples one by one from this tuple store.
+ */
+TupleTableSlot *
+RealTimeExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
```

Comment:
All of them looks same to me too; finished, performed, completed.

I will let you guys pick one :)